climada/engine/impact.py

```diff
@@ -940,6 +946,172 @@ def _cen_return_imp(imp, freq, imp_th, return_periods):
+
+        Raises
+        ------
+        ValueError
+            If the impact matrix is missing, the eai_exp and aai_agg cannot
+            be updated for a selection of events and/or exposures.
+
+        Returns
+        -------
+        imp : Impact()
+            A new impact object with a selection of events and/or exposures
+
+        """
+
+        if self.imp_mat.shape == (0, 0):
+            raise ValueError("The impact matrix is missing. eai_exp and\
+aai_agg cannot be computed. Recomputed impact.calc with save_mat=True")
+
+        imp = copy.deepcopy(self)
+        nb_events = imp.event_id.size
+        nb_exp = len(imp.coord_exp)
+        
+        sel_exp = np.arange(nb_exp)
+        # filter exposures
+        if isinstance(coord_exp, np.ndarray):
+            sel_exp = [j
+                       for j, coord in enumerate(imp.coord_exp)
+                       if coord in coord_exp
+                      ]
+
+        # filter events by date
+        if dates is None:
+            sel_dt = np.zeros(nb_events, dtype=bool)
+        else:
+            sel_dt = np.ones(nb_events, dtype=bool)
+            date_ini, date_end = dates
+            if isinstance(date_ini, str):
+                date_ini = util_dt.str_to_date(date_ini)
+                date_end = util_dt.str_to_date(date_end)
+            sel_dt &= (date_ini <= imp.date) & (imp.date <= date_end)
+            if not np.any(sel_dt):
+                LOGGER.info('No impact event in date range %s.', dates)
+                return None     
+        sel_dt = list(np.argwhere(sel_dt).reshape(-1))
+        
+        # filter events by id
+        if isinstance(event_ids, list):
+            sel_id = [list(imp.event_id).index(_id) for _id in event_ids]
+        else:
+            sel_id = []
+
+
+        # filter events by name
+        if isinstance(event_names, list):
+            sel_na= [list(imp.event_name).index(name) for name in event_names]
+        else:
+            sel_na = []
+
+        
+        #select events with machting id, name or date field.
+        sel_ev = [idx for idx in set(sel_dt + sel_id+  sel_na)]
+        
+        #if no date, name or id is set, take all events:
+        if (dates, event_ids, event_names) == (None, None, None):
+            sel_ev = np.arange(nb_events)
+
+        # set all attributes that are 'per event', i.e. have a row or a column
+        # of length equal to the number of events (=nb_events), using the 
+        # selection sel_ev.
+        per_event_attr = self.get_per_event_attr(imp)
+        for (attr, value) in imp.__dict__.items():
+            if attr in per_event_attr:
+                if isinstance(value, np.ndarray) and value.ndim == 1 \
+                                                   and value.size > 0:
+                    setattr(imp, attr, value[sel_ev])
+                elif isinstance(value, sparse.csr_matrix):
+                    setattr(imp, attr, value[sel_ev, :][:, sel_exp])
```

Comment:
`coord_exp` is only relevant for `sparse.csr_matrix` and not for `np.ndarray` then?
If that's correct it could perhaps be mentioned in the docstring.

Reply:
It is actually relevant also for the recomputations of the values imp.at_event, imp.eai_exp, and imp.aai_agg. It also sets imp.coord_ex.

Reply:
Really? What I can see now is the influence on `imp.coord_exp` in lines 1104 and 1105, but I can't see how those are changed. Can you point it out?

Reply:
I can see `imp.at_event`, `imp.coord_exp` and `imp.tot_value` below, true. (Though not yet `imp.eai_exp` or `imp.aai_agg`?)
But still - if `value` is an array, couldn't it have an exposure dimension as well?

Reply:
The `imp.eai_exp` and `imp.aai_agg` are computed from `imp_mat` which is set in line 1089 and which has indices select by `sel_exp`. Hence, if `sel_exp` is not None, the columns of `imp_mat` are modified, and then in turn `imp.eai_exp` and `imp.aai_agg` (in lines 1101, 1102).

Reply:
Furthermore, yes,  if `value` is an array, it could have an exposure dimension as well. However, due to what the impact is, it is rather unlikely that this happens. So far, I decided not to do the same treatment as for event dimensions. I fear that if further values with exposure dimension are added to the impact object, it will probably require a rewriting of the select method to make sure that the other impact values are correctly recomputed. Thus, we could hide a potential future issue here. But maybe I should rethink about that a bit more.

Reply:
I could rewrite the method `get_per_event_attr()` to be `get_per_attr(per_what)` and easily make the same logic for per event as for per events. Should I then place this new `get_per_attr(per_what)` function in util maybe?

Reply:
Why not make it a method of Impact?

Reply:
> 
> 
> ... However, due to what the impact is, it is rather unlikely that this happens. ...
Afaik "unlikely" is just the same as "granted" 😜

Reply:
> Why not make it a method of Impact?

It could be used in all the `select` methods which appear/will appear in different CLIMADA classes.

Reply:
> > ... However, due to what the impact is, it is rather unlikely that this happens. ...
> 
> Afaik "unlikely" is just the same as "granted" 😜

Yes, but this is part of the reflection. If it happens, then I suspect that the method would need some direct attention. But, I guess that the unit tests should take care of that.

Reply:
> 
> It could be used in all the `select` methods which appear/will appear in different CLIMADA classes.

Alright. In this case I suggest to call it by number, e.g.:
```
def get_attributes_with_matching_dimension(n: int):
    ...
    return list_of_attributes
```

Reply:
or even
```
def get_attributes_with_matching_dimension(*ns: int):
    ...
    return list_of_attributes # that have len(ns) dimensions or more, and all ns are individual part of shape
```

Reply:
So, I added the more general function to `util/select.py` (still need to write the test for that). 

I also slightly restructured the `impact.select` to use the more general attribute select function and added warnings when no events/exposures are found.

Reply:
Ok, all done.